include/ensmallen_bits/nsga2/nsga2.hpp

```diff
@@ -0,0 +1,313 @@
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef ENSMALLEN_NSGA2_NSGA2_HPP
+#define ENSMALLEN_NSGA2_NSGA2_HPP
+
+namespace ens {
+
+/**
+ * NSGA-II (Non-dominated Sorting Genetic Algorithm - II) is a multi-objective
+ * optimization algorithm. This class implements the NSGA-II algorithm.
+ *
+ * The algorithm works by generating a candidate population from a fixed
+ * starting point. At each stage of optimization, a new population of children
+ * is generated. This new population along with its predecessor is sorted using
+ * non-domination as the metric. Following this, the population is further
+ * segregated in fronts. A new population is generated from these fronts having
+ * size equal to that of the starting population.
+ *
+ * During evolution, two parents are randomly chosen using binary tournament
+ * selection. A pair of children are generated by crossing over these two
+ * candidates followed by mutation.
+ *
+ * The best front (Pareto optimal) is returned by the Optimize() method.
+ *
+ * For more information, see the following:
+ *
+ * @code
+ * @article{10.1109/4235.996017,
+ *   author = {Deb, K. and Pratap, A. and Agarwal, S. and Meyarivan, T.},
+ *   title = {A Fast and Elitist Multiobjective Genetic Algorithm: NSGA-II},
+ *   year = {2002},
+ *   url = {https://doi.org/10.1109/4235.996017},
+ *   journal = {Trans. Evol. Comp}}
+ * @endcode
+ *
+ * NSGA-II can optimize arbitrary multi-objective functions. For more details,
+ * see the documentation on function types included with this distribution or
+ * on the ensmallen website.
+ */
+class NSGA2 {
+ public:
+  /**
+   * Constructor for the NSGA2 optimizer.
+   *
+   * The default values provided over here are not necessarily suitable for a
+   * given function. Therefore it is highly recommended to adjust the
+   * parameters according to the problem.
+   *
+   * @param populationSize The number of candidates in the population.
+   *     This should be atleast 4 in size and a multiple of 4.
+   * @param maxGenerations The maximum number of generations allowed for NSGA-II.
+   * @param crossoverProb The probability that a crossover will occur.
+   * @param mutationProb The probability that a mutation will occur.
+   * @param mutationStrength The strength of the mutation.
+   * @param epsilon The minimum difference required to distinguish between
+   *     candidate solutions.
+   * @param lowerBound Lower bound of the coordinates of the initial population.
+   * @param upperBound Upper bound of the coordinates of the initial population.
+   */
+  NSGA2(const size_t populationSize = 100,
+        const size_t maxGenerations = 2000,
+        const double crossoverProb = 0.6,
+        const double mutationProb = 0.3,
+        const double mutationStrength = 1e-3,
+        const double epsilon = 1e-6,
+        const arma::vec lowerBound = arma::ones(1, 1),
```

Comment:
Maybe it make sense to use: `const arma::vec& lowerBound` here to be consistent with the rest of the codebase.

Reply:
Alright, I am changing both the upper and lower bounds to `const arma::vec&`